designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,279 @@
+- Repo: eslint/eslint
+- Start Date: 2021-11-30
+- RFC PR: <https://github.com/eslint/rfcs/pull/84>
+- Authors: [bmish](https://github.com/bmish)
+
+# Stricter validation of rule tests
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+We will add additional assertions to the rule test runner to address several potential points of poor test coverage, especially related to [rule suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).
+
+Assertions added to invalid test case error objects:
+
+- Must contain `suggestions` if the test case produces suggestions
```

Comment:
@fisker I just want to capture your comment here:

https://github.com/eslint/eslint/issues/15104#issuecomment-1050945616

> 👎 for this.
> 
> In some rule, I only add one test for the `suggestions`, maybe because it's simple and good enough, then focus other tests on the auto-fix or something else, eg "message", "report loaction", why I need assert `suggestions` in all those?
> 
> The tester should not force anything, test one thing in one case is good practice!

I do think this is a fair point. However, you could make the same point about this existing requirement to test autofixes added in [ESLint v7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class):
> AssertionError [ERR_ASSERTION]: The rule fixed the code. Please add 'output' property.

My new requirement about testing suggestions is just an extension of the existing requirement about testing autofixes. The question is, should we have both requirements to be consistent and to ensure maximum test coverage, have only one because we consider one more important than the other, or have neither? I lean towards enforcing maximum test coverage, but open to discussing this.

Reply:
I think the damage done by forgetting to test a suggestion is greater than the annoyance of needing to copy-paste suggestions into multiple test cases. We want to ensure the highest quality for rules so I’m 👍 for enforcing suggestion testing by default.

I do wonder, though, if maybe there might be a future opportunity to allow devs to select which of the automatic tests and validations they might want to opt out of. I’m not convinced it’s needed now, just something to think about.

Reply:
Maybe we could allow `suggestions: <number>` as an opt-out from testing details of the suggestions. We already allow `errors: <number>`.

`suggestions: 1` would assert that there's 1 suggestion.

The absence of `suggestions` property would assert that there are no suggestions, as proposed in this RFC.

Reply:
I think that’s a fine compromise.

Reply:
I like the idea of `suggestions: <number>` as a compromise. And I'm fine to allow both the `errors: <number>` and `suggestions: <number>` shorthands in all situations. Updated.